addons/utm/models/utm_source.py

```diff
@@ -71,13 +71,35 @@ def create(self, vals_list):
 
         return super().create(vals_list)
 
-    def write(self, values):
-        if values.get(self._rec_name) and not values.get('name'):
-            values['name'] = self.env['utm.source']._generate_name(self, values[self._rec_name])
-        if values.get('name'):
-            values['name'] = self.env['utm.mixin']._get_unique_names(self._name, [values['name']])[0]
+    def find_smalles_missing(self, values):
+        num_set = set(values)
+        current = 2
+        while True:
+            if current not in num_set:
+                return current
+            current += 1
```

Comment:
A cool math trick is that you can compute the sequential sum from L to F with `(L - F + 1) * (F + L) * 1/2`.
So you could `sum(values)`, which normally very fast, and then check against that formula which is also fast.
As in most cases you won't be missing a number.
```suggestion
    def _find_smallest_missing(self, values):
        if not values:
            return None
        # check if any is missing
        if sum(values) == (max(values) + 1) * max(values) / 2:
            return max(values + 1)
        
        num_set = set(values)
        current = 2
        while current not in num_set:
            current += 1
         return current
```